### PR TITLE
EMP-2230, REQ-6895 capi changes in deployment and operations: Create CFT for CAPI EMP

### PR DESCRIPTION
**Pull request by Rohitrajak1807**
Ticket links: https://platform9.atlassian.net/browse/EMP-2230
https://platform9.atlassian.net/browse/REQ-6895

### CODE REVIEW

**Review by Shubhamag12 — APPROVED** [review]

File: emp/emp_user_cftemplate.yaml [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+          Resource:
+          - '*'
+        Version: 2012-10-17
+      Roles:
+      - Ref: AWSIAMRoleControlPlane
+    Type: AWS::IAM::ManagedPolicy
+  AWSIAMManagedPolicyCloudProviderNodes:
+    Properties:
+      Description: For the Kubernetes Cloud Provider AWS nodes
+      ManagedPolicyName: nodes.cluster-api-provider-aws.sigs.k8s.io
```
Comment (maheshbaviskar) [inline emp/emp_user_cftemplate.yaml:101]: @Rohitrajak1807 we need to fix this right ? or make it work even if existing object with same name exist.

Comment (Rohitrajak1807) [reply emp/emp_user_cftemplate.yaml:101]: Yes that needs change in controller, and then we'll have to change CFT

File: emp/emp_role_cftemplate.yaml [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+              - ecr:DescribeRepositories
+              - ecr:ListImages
+              - ecr:BatchGetImage
+            Effect: Allow
+            Resource:
+              - '*'
+          - Action:
+              # capa creates encrypted cloud init secrets, this is needed.
+              - secretsmanager:DeleteSecret
+              - secretsmanager:GetSecretValue
```
Comment (OmkarDeshpande7) [inline emp/emp_role_cftemplate.yaml:38]: We disable secrets manager. would this still be needed?

File: emp/emp_role_cftemplate.yaml [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+              - ec2:AcceptVpcPeeringConnection
+              - ec2:CreateVpcPeeringConnection
+              - ec2:DeleteVpcPeeringConnection
+              - ec2:DescribeVpcPeeringConnections
+              # these modify a route table, of a subnet. It's not possible to constrain the access from our side
+              # this is because, we use the VPCs already created in the customer's environment by the customer.
+              # the customer is free to add restrictions according to their use case. ec2:DescribeRouteTables, is also
+              # needed by emp-server to fill in the data for emp-ui, so the restrictions here must match the one defined
+              # below.
+              - ec2:DescribeRouteTables
```
Comment (OmkarDeshpande7) [inline emp/emp_role_cftemplate.yaml:267]: already in capa policies

File: emp/emp_role_cftemplate.yaml [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+              - '*'
+            Effect: Allow
+          - Action:
+              - iam:GetRole
+              - iam:GetRolePolicy
+              - iam:ListAttachedRolePolicies
+              - iam:ListRolePolicies
+              - iam:GetPolicy
+              - iam:GetPolicyVersion
+              - iam:SimulateCustomPolicy
```
Comment (OmkarDeshpande7) [inline emp/emp_role_cftemplate.yaml:286]: why is this needed?

File: emp/emp_role_cftemplate.yaml [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+              - ec2:DescribeVolumes
+              # needed by emp-server to populate the data on the UI. We can't dictate what resources customer will use therefore resource *
+              # the customer has full freedom to restrict access to limited resources, and we encourage to do it according to their use case.
+              - ec2:DescribeRouteTables
+              - ec2:DescribeVpcs
+              - ec2:DescribeRegions
+              - ec2:DescribeSecurityGroups
+              - ec2:DescribeSubnets
+              - ec2:DescribeKeyPairs
+              - ec2:DescribeAvailabilityZones
```
Comment (OmkarDeshpande7) [inline emp/emp_role_cftemplate.yaml:330]: Most of them taken care in CAPA permissions

**Review by OmkarDeshpande7 — COMMENTED** [review]
Looks good

File: emp/emp_role_cftemplate.yaml (hunk 2 above)
Comment (Rohitrajak1807) [reply emp/emp_role_cftemplate.yaml:38]: yes, we're going to re-enable them after 1.1

File: emp/emp_role_cftemplate.yaml (hunk 3 above)
Comment (Rohitrajak1807) [reply emp/emp_role_cftemplate.yaml:267]: yes, putting it here was intentional. these permissions for now are grouped according to the features or components that need it. It's easier to track it like this.

File: emp/emp_role_cftemplate.yaml (hunk 4 above)
Comment (Rohitrajak1807) [reply emp/emp_role_cftemplate.yaml:286]: for `iam:SimulateCustomPolicy` when it's reworked

File: emp/emp_role_cftemplate.yaml (hunk 5 above)
Comment (Rohitrajak1807) [reply emp/emp_role_cftemplate.yaml:330]: see: https://github.com/platform9/support-locker/pull/50#discussion_r1743087737
putting it here is intentional. They'll be removed once we have tagging in place and know for sure the exact scope of each permission

**Comment by Rohitrajak1807** [comment]
Holding this, as we remove unneeded permissions.

**Comment by Rohitrajak1807** [comment]
> Holding this, as we remove unneeded permissions.

done

**Review by Shubhamag12 — APPROVED** [review]

File: emp/emp_role_cftemplate.yaml [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+            Resource:
+              - arn:*:secretsmanager:*:*:secret:aws.cluster.x-k8s.io/*
+          - Action:
+              # they are related to some heartbeat sent by systems manager see: https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-setting-up-messageAPIs.html
+              - ssm:UpdateInstanceInformation
+              - ssmmessages:CreateControlChannel
+              - ssmmessages:CreateDataChannel
+              - ssmmessages:OpenControlChannel
+              - ssmmessages:OpenDataChannel
+              # discuss why this is needed
```
Comment (AnirudhPokala) [inline emp/emp_role_cftemplate.yaml:50]: need to remove?

File: emp/emp_role_cftemplate.yaml [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+              - '*'
+          - Action:
+              # capa creates encrypted cloud init secrets, this is needed.
+              - secretsmanager:DeleteSecret
+              - secretsmanager:GetSecretValue
+            Effect: Allow
+            Resource:
+              - arn:*:secretsmanager:*:*:secret:aws.cluster.x-k8s.io/*
+          - Action:
+              # they are related to some heartbeat sent by systems manager see: https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-setting-up-messageAPIs.html
```
Comment (AnirudhPokala) [inline emp/emp_role_cftemplate.yaml:44]:
```suggestion
              # they are related to heartbeat sent by systems manager see: https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-setting-up-messageAPIs.html
```

File: emp/emp_role_cftemplate.yaml [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+              - elasticloadbalancing:RemoveTags
+              - elasticloadbalancing:SetSubnets
+              - elasticloadbalancing:ModifyTargetGroupAttributes
+              - elasticloadbalancing:CreateTargetGroup
+              - elasticloadbalancing:DescribeListeners
+              - elasticloadbalancing:CreateListener
+              - elasticloadbalancing:DescribeTargetHealth
+              - elasticloadbalancing:RegisterTargets
+              - elasticloadbalancing:DeleteListener
+              # commenting out capa actions, reason: we don't use machinepools, and just scale the md
```
Comment (AnirudhPokala) [inline emp/emp_role_cftemplate.yaml:158]: do we want to remove this commented part? as user will just download and apply the cft.

Comment (AnirudhPokala) [reply emp/emp_role_cftemplate.yaml:158]: why do we need this commented parts in public exposed cft.

File: emp/emp_role_cftemplate.yaml (hunk 6 above)
Comment (Rohitrajak1807) [reply emp/emp_role_cftemplate.yaml:50]: its needed by capi :)

Comment (AnirudhPokala) [reply emp/emp_role_cftemplate.yaml:50]: OK, my only concern is as we put cft this to public folks, these commented actions will be confusing to users. As he is not aware CAPI or Qbert. He is only accounted for EMP Cloud formation template.

Comment (AnirudhPokala) [reply emp/emp_role_cftemplate.yaml:50]: Do we plan to update this policy before 1.1 again to remove all commented part and make it clean?

**Review by AnirudhPokala — APPROVED** [review]

**Review by khallnayak — APPROVED** [review]
